### PR TITLE
Fix and simplify error handling in (RSA/EC)_new_method()

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Idea comes from exchanges in PR #873.
Nearly same as 2bbf0baa9565cce937d85592f12d1279871dfaf6.

### CODE REVIEW

File: crypto/ec/ec_kmeth.c [hunk 1]
```diff
@@ -152,23 +145,23 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
...
     ret->conv_form = POINT_CONVERSION_UNCOMPRESSED;
-    ret->references = 1;
+
+    if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_EC_KEY, ret, &ret->ex_data)) {
+        goto err;
+    }
 
     if (ret->meth->init != NULL && ret->meth->init(ret) == 0) {
         ECerr(EC_F_EC_KEY_NEW_METHOD, ERR_R_INIT_FAIL);
+err:
```
Comment (mattcaswell) [inline crypto/ec/ec_kmeth.c]: I think we should avoid this...i.e. a label in the middle of an "if" that stuff outside the "if" can jump to. I'd prefer it if the label was at the end of the function and this "if" just does a "goto err".

File: crypto/rsa/rsa_lib.c [hunk 2]
```diff
@@ -141,34 +148,19 @@ RSA *RSA_new_method(ENGINE *engine)
...
-#endif
-        CRYPTO_free_ex_data(CRYPTO_EX_INDEX_RSA, ret, &ret->ex_data);
-        OPENSSL_free(ret);
-        return NULL;
+        goto err;
     }
 
     if ((ret->meth->init != NULL) && !ret->meth->init(ret)) {
+        RSAerr(RSA_F_RSA_NEW_METHOD, ERR_R_INIT_FAIL);
+err:
```
Comment (mattcaswell) [inline crypto/rsa/rsa_lib.c]: and here

**Comment by FdaSilvaYY** [comment]
Done. 
Let' wait for Travis results.

**Comment by mattcaswell** [comment]
+1

**Comment by richsalz** [comment]
+1; you or me merge?

**Comment by mattcaswell** [comment]
Merged. Thanks.

**Pull request closed without merging**